master/internal/rm/kubernetesrm/job.go

```diff
@@ -143,284 +140,363 @@ func newPod(
+		j.informTaskResourcesState()
 
 	case cproto.Running:
-		p.syslog.Infof("transitioning pod state from %s to %s", p.container.State, containerState)
-		p.container = p.container.Transition(cproto.Running)
-		p.informTaskResourcesStarted(getResourcesStartedForPod(p.pod, p.ports))
-		err := p.startPodLogStreamer()
-		if err != nil {
-			return p.container.State, err
-		}
+		j.syslog.Infof("saw pod state go from %s to %s", j.container.State, cproto.Running)
+		j.container.State = cproto.Running
+		j.informTaskResourcesStarted(sproto.ResourcesStarted{NativeResourcesID: j.jobName})
 
 	case cproto.Terminated:
-		exitCode, exitMessage, err := getExitCodeAndMessage(p.pod, p.containerNames)
+		if j.firstPodExit != nil {
+			return nil
+		}
+
+		exitCode, exitMessage, err := getExitCodeAndMessage(&updatedPod, j.containerNames)
 		if err != nil {
 			// When a pod is deleted, it is possible that it will exit before the
 			// determined containers generates an exit code. To check if this is
 			// the case we check if a deletion timestamp has been set.
-			if p.pod.ObjectMeta.DeletionTimestamp != nil {
-				p.syslog.Info("unable to get exit code for pod, setting exit code to 1025")
+			if j.job.ObjectMeta.DeletionTimestamp != nil {
+				j.syslog.Info("unable to get exit code for pod, setting exit code to 1025")
 				exitCode = 1025
 				exitMessage = "unable to get exit code or exit message from pod"
 			} else {
-				return p.container.State, err
+				return err
 			}
 		}
-
-		p.syslog.Infof("transitioning pod state from %s to %s", p.container.State, containerState)
-		p.container = p.container.Transition(cproto.Terminated)
-
-		var resourcesStopped sproto.ResourcesStopped
-		switch exitCode {
-		case aproto.SuccessExitCode:
-			p.syslog.Infof("pod exited successfully")
-		default:
-			p.syslog.Infof("pod failed with exit code: %d %s", exitCode, exitMessage)
-			resourcesStopped.Failure = sproto.NewResourcesFailure(
-				sproto.ResourcesFailed,
-				exitMessage,
-				ptrs.Ptr(sproto.ExitCode(exitCode)))
+		j.firstPodExit = &podExitInfo{
+			code: exitCode,
+			msg:  exitMessage,
 		}
-		p.informTaskResourcesStopped(resourcesStopped)
-		return p.container.State, nil
 
 	default:
-		panic(fmt.Sprintf("unexpected container state %s", containerState))
+		panic(fmt.Sprintf("unexpected container state %s", podState))
 	}
 
-	return p.container.State, nil
+	return nil
 }
 
-func (p *pod) podEventUpdate(event *k8sV1.Event) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+func (j *job) podDeleted(updatedPod *k8sV1.Pod) {
+	j.mu.Lock()
+	defer j.mu.Unlock()
 
-	// We only forward messages while pods are starting up.
-	switch p.container.State {
-	case cproto.Running, cproto.Terminated:
+	if j.firstPodExit != nil {
 		return
 	}
 
-	msgText := p.preparePodUpdateMessage(event.Message)
-	event.Message = msgText
+	j.firstPodExit = &podExitInfo{
+		code: 1027,
+		msg: fmt.Sprintf(
+			"pod %s was suddenly deleted but not by us (setting exit code 1027)",
```

Comment:
ditto, pull out and make consts so it's easier to find when debugging

Reply:
i dont agree that this makes it easier to debug. if it's in a const i grep 1027, find the const, and go to where it is used. if it is this way, i just find the code where it is used. and i dont think either is very different.